docs/guides/signals.md

```diff
@@ -0,0 +1,177 @@
+        signal_metadata: t.Dict[str, t.Union[str, int, float, bool]]
+    ):
+        self.signal_metadata = signal_metadata
+
+    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
+        threshold = self.signal_metadata["threshold"]
+        return random.random() > threshold
+```
+
+Note that the `RandomSignal` class sub-classes `Signal` and takes a `signal_metadata` argument.
+
+The `check_intervals()` method extracts the threshold metadata and compares a random number to it.
+
+We can now add a factory function that returns the `RandomSignal` sub-class. Note the `@signal_factory` decorator on line 8:
+
+```python linenums="1" hl_lines="8-10"
+import random
+import typing as t
+from sqlmesh.core.scheduler import signal_factory, Batch, Signal
+
+class RandomSignal(Signal):
+    def __init__(
+        self,
+        signal_metadata: t.Dict[str, t.Union[str, int, float, bool]]
+    ):
+        self.signal_metadata = signal_metadata
+
+    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
+        threshold = self.signal_metadata["threshold"]
+        return random.random() > threshold
+
+@signal_factory
+def my_signal_factory(signal_metadata: t.Dict[str, t.Union[str, int, float, bool]]) -> Signal:
+    return RandomSignal(signal_metadata)
+```
+
+We now have a working signal!
+
+We specify that a model should use the signal by passing metadata to the model DDL's `signals` key.
+
+The `signals` key accepts an array delimited by brackets `[]`. Each tuple in the list should contain the metadata needed for one signal evaluation.
+
+This example specifies that the `RandomSignal` should evaluate once with a threshold of 0.5:
+
+```sql linenums="1" hl_lines="4-6"
+MODEL (
+  name example.signal_model,
+  kind FULL,
+  signals [
+    (threshold = 0.5), # specify threshold value
+  ]
+);
+
+SELECT 1
+```
+
+The next time this project is `sqlmesh run`, our signal will metaphorically flip a coin to determine whether the model should be evaluated.
+
+### Advanced Example
+
+Multiple signals can be used within one model. A signal can return a subset of the intervals within a batch. The intersection of the intervals returned by all the signals determines what is evaluated.
+
+In this example, there are two signals.
+
+```python
+import typing as t
+from datetime import datetime
+
+from sqlmesh.core.scheduler import signal_factory, Batch, Signal
+from sqlmesh.utils.date import to_datetime
+
+
+class AlwaysReady(Signal):
+    # signal that indicates every interval is always ready
+    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
+        return True
+
+
+class OneweekAgo(Signal):
+    def __init__(self, dt: datetime):
+        self.dt = dt
+
+    # signal that returns only intervals that are <= 1 week ago
+    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
+        return [
+            (start, end)
+            for start, end in batch
+            if start <= self.dt
+        ]
+```
+
```

Comment:
```suggestion
Instead of returning a single `True`/`False` value for whether a batch of intervals is ready for evaluation, the `OneweekAgo` signal class returns specific intervals from the batch.

Its `check_intervals` method accepts a `dt` datetime argument, to which It compares the beginning of each interval in the batch. If the interval start is before that argument, the interval is ready for evaluation and included in the returned list.
```